### PR TITLE
Add instrumentation tests for saved form using espresso

### PR DESCRIPTION
**Pull request by ShivamPokhriyal**
Tests are running successfully using [jenkins and aws](https://jenkins.dimagi.com/job/commcare-android-instrumentation-tests/23/)

cross-request: https://github.com/dimagi/commcare-core/pull/901

[Jenkins setup](https://jenkins.dimagi.com/job/commcare-android-instrumentation-tests): 

- `assembleCommcareDebug` and `assembleCommcareDebugAndroidTest` gradle tasks are used to create 2 apks, 1 debug and 1 for testing. 
- Once the build is successful, we add a post build action to run tests on aws device farm. It uses jenkins aws device farm [plugin](https://github.com/jenkinsci/aws-device-farm-plugin/).  

[BrowserStack run](https://app-automate.browserstack.com/dashboard/v2/builds/79ac9df189fb4e9dc24beb3a2a19391c4048f689) It appears to me that browserstack is more powerful and stable compared to aws device farm. Though they don't have a jenkins plugin for espresso as of now. 

Added a python script which will run in [this](https://jenkins.dimagi.com/job/commcare-android-instrumentation-tests/) jenkins job and will schedule a build on browserstack and at the same time will run the tests on aws-device-farm. 

WIKI: https://confluence.dimagi.com/display/MD/CommCare+Android+Instrumentation+Tests

### CODE REVIEW

File: app/instrumentation-tests/src/org/commcare/tests/SavedFormTest.java [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+
+        // Stub out the Camera. When an intent is sent to the Camera, this tells Espresso to respond
+        // with the ActivityResult we just created
+        intending(hasAction(MediaStore.ACTION_IMAGE_CAPTURE)).respondWith(result);
+    }
+
+    @Before
+    public void login() {
+        if (CommCareApplication.instance().getCurrentApp() == null) {
+            Utility.installApp("2zI4jeQ");
```
Comment (shubham1g5) [inline app/instrumentation-tests/src/org/commcare/tests/SavedFormTest.java:101]: can we offline install this from ccz in order to have faster test runtime ?

File: app/src/org/commcare/tasks/SaveToDiskTask.java [hunk 2]
```diff
@@ -289,25 +290,21 @@ public void setFormSavedListener(FormSavedListener fsl) {
      * though, until all answers conform to their constraints/requirements.
      */
     private boolean hasInvalidAnswers(boolean markCompleted) {
-        FormIndex i = FormEntryActivity.mFormController.getFormIndex();
-        FormEntryActivity.mFormController.jumpToIndex(FormIndex.createBeginningOfFormIndex());
-
+        FormController formController = FormEntryActivity.mFormController;
```
Comment (shubham1g5) [inline app/src/org/commcare/tasks/SaveToDiskTask.java:293]: Is there a reason these changes are needed ?

File: app/src/org/commcare/activities/FormEntryActivity.java [hunk 3]
```diff
@@ -257,8 +257,9 @@ public void onReceive(Context context, Intent intent) {
     @Override
     protected void onSaveInstanceState(Bundle outState) {
         super.onSaveInstanceState(outState);
-        instanceState.saveState(outState);
-
+        if (instanceState != null) {
```
Comment (shubham1g5) [inline app/src/org/commcare/activities/FormEntryActivity.java:260]: What scenarios lead to `instanceState` being null ?

Comment (ShivamPokhriyal) [reply app/src/org/commcare/activities/FormEntryActivity.java:260]: I honestly have no idea why it would ever be null here. I thought that this might be a result of using IntentTestRule [here](https://github.com/dimagi/commcare-android/pull/2231/files#diff-14085b87043ea3b3b0ad6a0f2be4f793R59) since the crash started happening after adding that only and it's needed to stub intents. 

But then I also saw some crashes related to this in [crashlytics](https://console.firebase.google.com/u/0/project/commcare-a57e4/crashlytics/app/android:org.commcare.dalvik/issues/d3f3ff13c96e62cb358bc94c1fc8fc99?time=last-seven-days&sessionId=5EC12E04004C00015A3462220FB79688_DNE_0_v2) as well.

File: app/src/org/commcare/tasks/SaveToDiskTask.java (hunk 2 above)
Comment (ShivamPokhriyal) [reply app/src/org/commcare/tasks/SaveToDiskTask.java:293]: Yeah, so the earlier code used to modify the `currentFormIndex` to loop over all the questions. But if at the same time, some other thread also wants to get hold of the current question in screen, it'll have to use the `currentFormIndex` which might give wrong value, if this thread has already modified it. And hence may produce NPE in worst cases. 

Since, we've to loop over all the questions here, we don't really need to modify the `currentFormIndex` rather we can simply have another form index to get all the questions. And this will also make sure that the other threads who might be using the `currentFormIndex` at the same time don't get wrong values.

File: app/src/org/commcare/activities/FormEntryActivity.java (hunk 3 above)
Comment (shubham1g5) [reply app/src/org/commcare/activities/FormEntryActivity.java:260]: Would be good to investigate this if you are encountering it in your test,  since doing this null check will only result in supressing errors we should not and can result in user believing everything is alright when it is not.

**Comment by shubham1g5** [comment]
@ShivamPokhriyal  Wondering if you will also be able to port couple other calabash tests to instrumentation tests as part of this or another PR, this will help us truly compare the two frameworks and will also set a solid foundation for instrumentation tests in our code base. Some tests that would be good candidate to try out will be [@CaseClaim](https://github.com/dimagi/commcare-ui-tests/blob/master/features/case_claim.feature) , @AsyncRestore (https://github.com/dimagi/commcare-ui-tests/blob/master/features/async_restore.feature)  and [@FormAttachmentUpload](https://github.com/dimagi/commcare-ui-tests/blob/master/features/form_attachment_upload.feature). These tests are a bit complex as they touch either device or HQ integration points and will be good for us to verify that Instrumentation tests can support.

Thanks again for piloting this, super excited about these!

File: app/src/org/commcare/tasks/SaveToDiskTask.java (hunk 2 above)
Comment (shubham1g5) [reply app/src/org/commcare/tasks/SaveToDiskTask.java:293]: Got it, thanks for the explanation. Only suggestion here would be to abstract this out into it's own PR so that it's easy to revert.

**Comment by ShivamPokhriyal** [comment]
@damagatchi retest this please.

**Comment by ShivamPokhriyal** [comment]
@damagatchi retest this please.

File: app/instrumentation-tests/src/org/commcare/utils/Utility.java [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+
+    private static void stubFileSelection(String filePath) {
+        Intent resultData = new Intent();
+        Uri fileUri = Uri.fromFile(new File(filePath));
+        resultData.setData(fileUri);
+        Instrumentation.ActivityResult result = new Instrumentation.ActivityResult(Activity.RESULT_OK, resultData);
+        intending(hasAction(Intent.ACTION_GET_CONTENT)).respondWith(result);
+    }
+
+    private static void onIntentSent(Intent intent) {
```
Comment (shubham1g5) [inline app/instrumentation-tests/src/org/commcare/utils/Utility.java:225]: how about `onImageCaptureIntentSent` ?

File: app/instrumentation-tests/src/org/commcare/utils/Utility.java [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+import static androidx.test.espresso.matcher.ViewMatchers.isRoot;
+import static androidx.test.espresso.matcher.ViewMatchers.withId;
+import static androidx.test.espresso.matcher.ViewMatchers.withText;
+import static org.hamcrest.Matchers.anything;
+import static org.hamcrest.Matchers.startsWith;
+
+/**
+ * @author $|-|!˅@M
+ */
+public class Utility {
```
Comment (shubham1g5) [inline app/instrumentation-tests/src/org/commcare/utils/Utility.java:47]: `InstrumentationUtility` ?

File: scripts/instrumentationRunner.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
+"""
```
[the remark this answers is not in the document]
Comment (ShivamPokhriyal) [reply scripts/instrumentationRunner.py:1]: This is for running the tests on browserstack. When I was working on it, no jenkins plugin was available for browserstack.

File: app/instrumentation-tests/src/org/commcare/utils/DateTime.java [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+ * href="http://tools.ietf.org/html/rfc3339">RFC 3339</a>.
+ *
+ * <p>
+ * Implementation is immutable and therefore thread-safe.
+ * </p>
+ *
+ * @since 1.0
+ * @author Yaniv Inbar
+ */
+public final class DateTime implements Serializable {
```
[the remark this answers is not in the document]
Comment (ShivamPokhriyal) [reply app/instrumentation-tests/src/org/commcare/utils/DateTime.java:37]: I remember trying the Android DateTime methods for this but couldn't get them to work as expected. So I used this utility instead. It was maybe the `Z` offset that caused some trouble to me.

File: app/instrumentation-tests/src/org/commcare/tests/CaseClaimTest.java [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+                1
+        )).perform(typeText(name));
+        closeSoftKeyboard();
+        onView(withId(R.id.request_button))
+                .perform(click());
+        onView(withId(R.id.request_button))
+                .check(matches(isDisplayed())); // The request doesn't do anything which confirms case doesn't exists.
+
+        // Register a new case with name cordelia
+        pressBackUnconditionally();
```
[the remark this answers is not in the document]
Comment (ShivamPokhriyal) [reply app/instrumentation-tests/src/org/commcare/tests/CaseClaimTest.java:73]: Yeah, I had a silly thought where I liked that pressBackUnconditionally is better since it doesn't throw any exceptions :)

File: app/instrumentation-tests/src/org/commcare/tests/CaseClaimTest.java [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+
+        onView(isRoot()).perform(Utility.sleep(TimeUnit.SECONDS.toMillis(15)));
+
+        Utility.login("claim_test2", "123");
+        // Syncing once again to pull all the case data.
+        onView(withText("Sync with Server"))
+                .perform(click());
+
+        // Somehow, espresso starts before data pull is completed.
+        onView(isRoot()).perform(Utility.sleep(TimeUnit.SECONDS.toMillis(15)));
```
[the remark this answers is not in the document]
Comment (ShivamPokhriyal) [reply app/instrumentation-tests/src/org/commcare/tests/CaseClaimTest.java:117]: I guess it's not possible to perform anything with progress bar in espresso.

File: app/instrumentation-tests/src/org/commcare/tests/CaseClaimTest.java [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+                .perform(click());
+        // Make sure espresso doesn't start before data pull is completed.
+        onView(isRoot()).perform(Utility.sleep(TimeUnit.SECONDS.toMillis(15)));
+
+        // Make sure the case is no longer around.
+        onView(withText("Start"))
+                .perform(click());
+        onView(withText("Follow Up"))
+                .perform(click());
+        onView(withText("SEARCH ALL CASES"))
```
[the remark this answers is not in the document]
Comment (ShivamPokhriyal) [reply app/instrumentation-tests/src/org/commcare/tests/CaseClaimTest.java:176]: I tried to have almost exactly the same flow as calabash. And I think we're doing the same there.

File: app/instrumentation-tests/src/org/commcare/tests/BaseTest.java [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+
+    protected void installApp(String appName, String ccz) {
+        if (CommCareApplication.instance().getCurrentApp() == null) {
+            Utility.installApp(ccz);
+        } else if (!appName.equals(CommCareApplication.instance().getCurrentApp().getAppRecord().getDisplayName())) {
+            // We already have an installed app, But not the one we need for this test.
+            Utility.uninstallCurrentApp();
+            Utility.installApp(ccz);
+            // App installation doesn't take back to login screen. Is this an issue?
+            pressBackUnconditionally();
```
[the remark this answers is not in the document]
Comment (ShivamPokhriyal) [reply app/instrumentation-tests/src/org/commcare/tests/BaseTest.java:52]: Yeah, I found it weird when I saw it first, but it's true!

File: app/instrumentation-tests/src/org/commcare/tests/BaseTest.java [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+ */
+@RunWith(AndroidJUnit4.class)
+@LargeTest
+public abstract class BaseTest {
+
+    @Rule
+    public ActivityTestRule<DispatchActivity> activityTestRule = new ActivityTestRule<>(DispatchActivity.class);
+
+    @Rule
+    public IntentsTestRule<FormEntryActivity> intentsRule = new IntentsTestRule<>(FormEntryActivity.class);
```
[the remark this answers is not in the document]
Comment (ShivamPokhriyal) [reply app/instrumentation-tests/src/org/commcare/tests/BaseTest.java:30]: Yeah I thought that most of our intents are from the FormEntryActivity only so decided to put it here. But yeah makes sense to avoid having it in a superclass when not all subclasses need it.

File: app/instrumentation-tests/src/org/commcare/AsyncRestoreHelperMock.java [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+
+import org.commcare.network.RemoteDataPullResponse;
+import org.commcare.tasks.AsyncRestoreHelper;
+import org.commcare.tasks.DataPullTask;
+import org.commcare.tasks.ResultAndError;
+
+/**
+ * @author $|-|!˅@M
+ */
+public class AsyncRestoreHelperMock extends AsyncRestoreHelper {
```
[the remark this answers is not in the document]
Comment (ShivamPokhriyal) [reply app/instrumentation-tests/src/org/commcare/AsyncRestoreHelperMock.java:11]: I think the reason why I couldn't use mockito is because I have no way to inject the mock [here](https://github.com/dimagi/commcare-android/pull/2231/files#diff-5ce2bf85747c398fb9c88402a8a693b2L109)

File: app/instrumentation-tests/src/org/commcare/tests/CaseClaimTest.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (shubham1g5) [reply app/instrumentation-tests/src/org/commcare/tests/CaseClaimTest.java:117]: It's weird in that case that espresso starts even before the data pull as seems like it waits by itself for the async tasks to be executed which is the reason we can't perform operations on progress bar right ? so both these behaviours looks contradicting to me with possibility of only one to exist. If they are both co-existing it would be nice to see what exactly is going on here.

File: app/instrumentation-tests/src/org/commcare/tests/CaseClaimTest.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (shubham1g5) [reply app/instrumentation-tests/src/org/commcare/tests/CaseClaimTest.java:176]: I can't find the "search all cases" [here](https://github.com/dimagi/commcare-ui-tests/blob/master/features/case_claim.feature#L102 ) as well.  Am I missing something ?

File: app/instrumentation-tests/src/org/commcare/utils/DateTime.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (shubham1g5) [reply app/instrumentation-tests/src/org/commcare/utils/DateTime.java:37]: Would you be able to have a relook and elaborate the issue you had with the `Z` offset.

File: scripts/instrumentationRunner.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (shubham1g5) [reply scripts/instrumentationRunner.py:1]: And a Jenkins plugin is available now ? I am mostly curious if there is a place where we are referencing this script or is browserstack auto-configured to look for this file.

[the remark this answers is not in the document]
Comment (ShivamPokhriyal) [reply scripts/instrumentationRunner.py:1]: This is used in jenkins job [here](https://jenkins.dimagi.com/job/commcare-android-instrumentation-tests/configure) at the execute shell option.

File: app/src/org/commcare/activities/FormEntryActivity.java (hunk 3 above)
Comment (ShivamPokhriyal) [reply app/src/org/commcare/activities/FormEntryActivity.java:260]: will remove this change from the PR before merging.

File: app/instrumentation-tests/src/org/commcare/tests/CaseClaimTest.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (ShivamPokhriyal) [reply app/instrumentation-tests/src/org/commcare/tests/CaseClaimTest.java]: updated. Sorry the confusion.

File: app/instrumentation-tests/src/org/commcare/tests/CaseClaimTest.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (ShivamPokhriyal) [reply app/instrumentation-tests/src/org/commcare/tests/CaseClaimTest.java]: yeah I confused myself and added 2 waits there. I just wanted to wait for HQ since it gave me wrong data many times. Updated the code to remove these waits.

File: app/instrumentation-tests/src/org/commcare/tests/BaseTest.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (ShivamPokhriyal) [reply app/instrumentation-tests/src/org/commcare/tests/BaseTest.java]: Now when I was updating it, I got to know that we need atleast 1 IntentsTestRule to be able to install the app offline. So keeping it here.

**Comment by ShivamPokhriyal** [comment]
@shubham1g5 Another review please :)

File: scripts/instrumentationRunner.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (shubham1g5) [reply scripts/instrumentationRunner.py:1]: cool, thanks.

File: app/instrumentation-tests/src/org/commcare/tests/BaseTest.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (shubham1g5) [reply app/instrumentation-tests/src/org/commcare/tests/BaseTest.java]: How is a intent test rule with `FormEntryActivity` related to offline ccz install ?

File: app/instrumentation-tests/src/org/commcare/AsyncRestoreHelperMock.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (shubham1g5) [reply app/instrumentation-tests/src/org/commcare/AsyncRestoreHelperMock.java:11]: is there a way we can mock the `DataPullTask` itself instead of `AsyncRestoreHelper` to check for this behaviour ? `AsyncRestoreHelper` is in an internal api to `DataPullTask` and should ideally be treated as a black box.

File: app/instrumentation-tests/src/org/commcare/tests/CaseClaimTest.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (shubham1g5) [reply app/instrumentation-tests/src/org/commcare/tests/CaseClaimTest.java]: sounds good.

File: app/instrumentation-tests/src/org/commcare/tests/BaseTest.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (ShivamPokhriyal) [reply app/instrumentation-tests/src/org/commcare/tests/BaseTest.java]: Ahh, I see it now. Actually we only need `IntentTestRule` with any activity, so will update the code. I used `ActivityTestRule` while starting to write tests and forgot to update that.

File: app/instrumentation-tests/src/org/commcare/AsyncRestoreHelperMock.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (ShivamPokhriyal) [reply app/instrumentation-tests/src/org/commcare/AsyncRestoreHelperMock.java:11]: Ohh, so it'll be like: I mock `DataPulLtask` and then override its constructor to use `AsyncRestoreHelperMock`. And then perhaps maybe I'll have to mock `FormAndDataSyncer` as well and override it's method to use `DataPullTaskMock`? Or maybe I'm wrong here?

File: app/instrumentation-tests/src/org/commcare/utils/DateTime.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (ShivamPokhriyal) [reply app/instrumentation-tests/src/org/commcare/utils/DateTime.java]: Putting it in here in case you're interested. The date string coming from HQ was like `2020-06-12T02:42:38.555065Z`
The closest I got to converting that to milliseconds was using `SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSsss'Z'")` and it gave `1591929785555` which is different than the expected `1591929758555`.
`java.time.Instant.parse()` gives correct result but it requires api level 26+ so can't use that. 
So, I'm using `org.joda.DateTime` right now.

File: scripts/instrumentationRunner.py [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+Submits the apks to BrowserStack to run the instrumentation tests.
```
Comment (shubham1g5) [inline scripts/instrumentationRunner.py:2]: since it's only related to browserstack we should rename this something similar to `browserstack.py` to avoid confusion.

File: scripts/instrumentationRunner.py [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+import sys
+import json
+
+def appendData(command, dataUrl):
+    var = "file=@{}".format(dataUrl)
+    return command + " " + json.dumps(var)
+
+def buildTestCommand(appToken, testToken):
+    test = {}
+    test["devices"] = ["LG G5-6.0"] #using just 1 device right now.
```
Comment (shubham1g5) [inline scripts/instrumentationRunner.py:18]: can we add another Android 4.4 device here ?

File: app/instrumentation-tests/src/org/commcare/AsyncRestoreHelperMock.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (shubham1g5) [reply app/instrumentation-tests/src/org/commcare/AsyncRestoreHelperMock.java:11]: Yeah you can make 2 changes - 
1.  Change `FormAndDataSyncer` to utilize a `DataPullTaskFactory` and then override that factory class in the tests to return you the mocked class. 
2. Have a setter for `asyncRestoreHelper` in `DataPullTask` to set the mocked `asyncRestoreHelper`

File: README.md [hunk 16]
```diff
@@ -92,6 +92,51 @@ Create a new Android Studio JUnit Build configuration using the following steps.
...
+
+```bash
+./gradlew connectedCommcareDebugAndroidTest -Pandroid.testInstrumentationRunnerArguments.class=<FULLY_QUALIFIED_NAME_OF_YOUR_TEST>
+```
+
+You can view the results from the output file generated.
+
+### Run instrumentation-tests from Android Studio
+
+Before running tests from Android-Studio make sure you've disabled animations in your device. Note, this is only required when you're running tests from Android Studio
```
[the remark this answers is not in the document]
Comment (shubham1g5) [reply README.md:121]: bumping this for your thoughts on this.

[the remark this answers is not in the document]
Comment (ShivamPokhriyal) [reply README.md:121]: Actually this is recommended from [espresso docs](https://developer.android.com/training/testing/espresso/setup?authuser=1#set-up-environment), I'm already using [animationDisabled](https://github.com/dimagi/commcare-android/pull/2231/files#diff-39e7d8c00954e920b98e7636f0ac30b2R226) gradle option for this but that doesn't work when you run tests using android studio.

[the remark this answers is not in the document]
Comment (shubham1g5) [reply README.md:121]: I see, and browserstack have these turn off by default ?

File: app/instrumentation-tests/src/org/commcare/AsyncRestoreHelperMock.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (shubham1g5) [reply app/instrumentation-tests/src/org/commcare/AsyncRestoreHelperMock.java:11]: Also realized you can also just introduce a factory for `asyncRestoreHelper` and override that in tests as well, Basically I am trying to avoid having any test specific functions in non-test code.

File: README.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (ShivamPokhriyal) [reply README.md:121]: No, the animationDisabled option that i used in the build.gradle file does the trick there I guess.

**Comment by ShivamPokhriyal** [comment]
@shubham1g5 One more round. 😃

File: app/instrumentation-tests/src/org/commcare/utils/DateTime.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (shubham1g5) [reply app/instrumentation-tests/src/org/commcare/utils/DateTime.java]: looks good.

**Comment by shubham1g5** [comment]
@ShivamPokhriyal  last set of changes look good. The only thing remaining is the `instanceState != null` check now.

**Comment by ShivamPokhriyal** [comment]
> @ShivamPokhriyal last set of changes look good. The only thing remaining is the `instanceState != null` check now.

Yup, gonna remove that before merging. https://github.com/dimagi/commcare-core/pull/901/ this supporting one needs a review as well.

File: scripts/browserstack.py [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+        userName = os.environ["BROWSERSTACK_USERNAME"]
+
+    if "BROWSERSTACK_PASSWORD" in os.environ:
+        password = os.environ["BROWSERSTACK_PASSWORD"]
+
+    if "BASE_LOCATION" in os.environ:
+        baseLoc = os.environ["BASE_LOCATION"]
+
+    debugApk = baseLoc + "commcare/debug/app-commcare-debug.apk"
+    testApk = baseLoc + "androidTest/commcare/debug/app-commcare-debug-androidTest.apk"
```
Comment (shubham1g5) [inline scripts/browserstack.py:36]: Q: Does brwserstack support uploading bundles ?

Comment (ShivamPokhriyal) [reply scripts/browserstack.py:36]: yeah, I noticed that sometime ago, seems like that's a new feature there. I'll change this to use app bundles instead.

Comment (ShivamPokhriyal) [reply scripts/browserstack.py]: Updated it to use app bundles. Also, the tests will be failing since I removed that null check! So until that one is fixed, they'll all fail!

Comment (ShivamPokhriyal) [reply scripts/browserstack.py]: But Imma gonna address that in a separate PR cuz that issue is coming to real users as well. So I'm good to merge this one unless you have any concerns.

File: scripts/browserstack.py [hunk 18]
```diff
@@ -32,8 +32,8 @@ def buildTestCommand(appToken, testToken):
     if "BASE_LOCATION" in os.environ:
         baseLoc = os.environ["BASE_LOCATION"]
 
-    debugApk = baseLoc + "commcare/debug/app-commcare-debug.apk"
-    testApk = baseLoc + "androidTest/commcare/debug/app-commcare-debug-androidTest.apk"
+    debugApk = baseLoc + "bundle/commcareDebug/app-commcare-debug.aab"
```
Comment (shubham1g5) [inline scripts/browserstack.py:35]: would be good to rename this to debugBundle now that it represents bundle.

File: scripts/browserstack.py (hunk 17 above)
Comment (shubham1g5) [reply scripts/browserstack.py]: @ShivamPokhriyal  I don't think there is value in us merging tests which are going to fail for time being. So unless you think otherwise, I will hold them off until the issue is fixed, verify that this PR works successfully on top of that other PR that fixes this isse and merge it afterwards.

Comment (ShivamPokhriyal) [reply scripts/browserstack.py]: Cool makes sense.

Comment (ShivamPokhriyal) [reply scripts/browserstack.py]: @shubham1g5 Hey I was wrong about it earlier. The tests are passing. The `IntentTestRule` was the reason behind the npe crashes and when we removed it here https://github.com/dimagi/commcare-android/pull/2231#discussion_r450157788, crashes also disappear.

Comment (shubham1g5) [reply scripts/browserstack.py]: wow, that's great.

**Comment by ShivamPokhriyal** [comment]
@damagatchi retest this please

**Comment by ShivamPokhriyal** [comment]
A perfect [build](https://jenkins.dimagi.com/job/commcare-android-instrumentation-tests/104)

I'm getting a lot of strict mode failures for `firebase-iid` in the instrumentation testing so tests are flaky right now. And I'll hold off merging this one until the `firebase-iid` is fixed.

**Comment by ShivamPokhriyal** [comment]
@damagatchi retest this please.

**Comment by ShivamPokhriyal** [comment]
@shubham1g5 Another review please. I've configured the espresso tests to run on release builds(which also avoids the StrictMode flakiness).
